pkg/frontend/openshiftcluster_preflightvalidation.go

```diff
@@ -0,0 +1,136 @@
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/frontend/middleware"
+)
+
+// /subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{resourceProviderNamespace}/deployments/{deploymentName}/preflight?api-version={api-version}
+func (f *frontend) preflightValidation(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	log := ctx.Value(middleware.ContextKeyLog).(*logrus.Entry)
+
+	var header http.Header
+	var b []byte
+
+	body := r.Context().Value(middleware.ContextKeyBody).([]byte)
+	correlationData := r.Context().Value(middleware.ContextKeyCorrelationData).(*api.CorrelationData)
+	systemData, _ := r.Context().Value(middleware.ContextKeySystemData).(*api.SystemData)
+	originalPath := r.Context().Value(middleware.ContextKeyOriginalPath).(string)
+	referer := r.Header.Get("Referer")
+
+	subId := chi.URLParam(r, "subscriptionId")
+	resourceGroupName := chi.URLParam(r, "resourceGroupName")
+
+	resources, err := unmarshalRequest(body)
+	if err != nil {
+		log.Warningf("bad Request. The request content was invalid and could not be deserialized: %s.", err)
+	} else {
+		resourceCount := len(resources.Resources)
+		ch := make(chan api.ValidationResult, resourceCount)
+		for _, raw := range resources.Resources {
+			// get typeMeta from the raw data
+			typeMeta := &api.ResourceTypeMeta{}
+			err = json.Unmarshal(raw, &typeMeta)
+			if err != nil {
+				// failing to parse the preflight body is not considered a validation failure. continue
+				log.Warningf("bad request. Failed to unmarshal ResourceTypeMeta: %s", err)
+				continue
+			}
+			if strings.EqualFold(typeMeta.Type, "Microsoft.RedHatOpenShift/openShiftClusters") {
+				path := PreflightResourceId(subId, resourceGroupName, typeMeta.Name)
+				ch <- f._preflightValidation(ctx, log, raw, correlationData, systemData, path, originalPath, r.Method, referer, &header, f.apis[typeMeta.APIVersion].OpenShiftClusterConverter, f.apis[typeMeta.APIVersion].OpenShiftClusterStaticValidator, subId, typeMeta.APIVersion, resourceGroupName)
+			}
+		}
+		close(ch)
+
+		for res := range ch {
+			// go through and serialize into an array and make that the return body
+			if res.Status == api.ValidationStatusFailed {
+				log.Warningf("preflight validation failed with: %q", res.Error)
+				resultByte, err := json.Marshal(res.Error)
+				if err != nil {
+					log.Warningf("the response could not be serialized: %s", err)
+				} else {
+					b = append(b, resultByte...)
+				}
+			}
+		}
+	}
+	reply(log, w, header, b, nil)
+}
+
+func (f *frontend) _preflightValidation(ctx context.Context, log *logrus.Entry, raw json.RawMessage, correlationData *api.CorrelationData, systemData *api.SystemData, path, originalPath, method, referer string, header *http.Header, converter api.OpenShiftClusterConverter, staticValidator api.OpenShiftClusterStaticValidator, subId, apiVersion string, resourceGroupName string) api.ValidationResult {
```

Comment:
In general, it's a bad practice to have functions with tons of arguments. This makes these functions (1) tightly coupled and (2) harder to test. Would it be possible to organize some of these arguments into a struct so we can just past the struct?

Reply:
it's more like an existing (or established) coding style. i'd suggest deferring this till later